src/tpm2-tss-engine-digest-sign.c

```diff
@@ -0,0 +1,304 @@
+                 size_t sig_size)
+{
+    TSS2_RC r;
+
+    if (!tpm2data)
+        /* non-TPM key - nothing to do */
+        return 1;
+
+    TPM2_SIG_DATA *data = calloc(sizeof(*data), 1);
+    if (!data) {
+        ERR(digest_sign_init, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    data->seq_handle = ESYS_TR_NONE;
+    data->sig_size = sig_size;
+
+    data->key = calloc(sizeof(*data->key), 1);
+    if (!data->key) {
+        ERR(digest_sign_init, ERR_R_MALLOC_FAILURE);
+        goto error;
+    }
+
+    data->key->refcount = 1;
+
+    r = init_tpm_key(&data->key->esys_ctx, &data->key->key_handle, tpm2data);
+    ERRchktss(digest_sign_init, r, goto error);
+    data->key->privatetype = tpm2data->privatetype;
+
+    EVP_PKEY_CTX_set_app_data(ctx, data);
+    /*
+     * Override the update function so that the TPM performs the
+     * digest, which is required for restricted keys - the TPM will
+     * reject a null validation ticket in this case for the signing
+     * operation.
+     */
+    EVP_MD_CTX_set_update_fn(mctx, digest_update);
+
+    if (!digest_init(mctx, data))
+        goto error;
+
+    return 1;
+
+ error:
+    if (data->key) {
+        if (data->key->key_handle != ESYS_TR_NONE) {
+            if (data->key->privatetype == KEY_TYPE_HANDLE) {
+                Esys_TR_Close(data->key->esys_ctx, &data->key->key_handle);
+            } else {
+                Esys_FlushContext(data->key->esys_ctx, data->key->key_handle);
+            }
+        }
+        if (data->key->esys_ctx)
+            esys_ctx_free(&data->key->esys_ctx);
+        free(data->key);
+    }
+    free(data);
+    return 0;
+}
+
+/**
+ * Copy digest and sign context
+ *
+ * @param dst Destination OpenSSL pkey context
+ * @param src Source OpenSSL pkey context
+ * @retval 1 on success
+ * @retval 0 on failure
+ */
+int
+digest_sign_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
+{
+    TPM2_SIG_DATA *src_sig_data = EVP_PKEY_CTX_get_app_data(src);
+    TPMS_CONTEXT *context = NULL;
+    TPM2_SIG_DATA *dst_sig_data = NULL;
+    TSS2_RC r;
+
+    if (src_sig_data) {
+        dst_sig_data = calloc(sizeof(*dst_sig_data), 1);
+        if (!dst_sig_data) {
+            ERR(digest_sign_copy, ERR_R_MALLOC_FAILURE);
+            return 0;
+        }
+
+        dst_sig_data->hash_alg = src_sig_data->hash_alg;
+
+        if (src_sig_data->seq_handle != ESYS_TR_NONE) {
+            /* duplicate sequence handle */
+
+            r = Esys_ContextSave(src_sig_data->key->esys_ctx,
+                                 src_sig_data->seq_handle, &context);
+            ERRchktss(digest_sign_copy, r, goto error);
+            dst_sig_data->seq_handle = ESYS_TR_NONE;
+            r = Esys_ContextLoad(src_sig_data->key->esys_ctx, context,
+                                 &dst_sig_data->seq_handle);
+            ERRchktss(digest_sign_copy, r, goto error);
+        }
+
+        dst_sig_data->key = src_sig_data->key;
+        atomic_fetch_add(&dst_sig_data->key->refcount, 1);
```

Comment:
Since we already have to contextsave/load the sequence handle to copy it for another context, cant we just dup the whole thing and avoid refcounting? Or could we use __atomic_add_fetch since gcc and clang support this since way back? Or perhaps try for atomic.h and if not, fall back to using builtins.

Reply:
My first approach was actually to dup the whole thing and not do the refcounting, but this causes a failure in one of the integration tests due to now running out of transient object handles. I would imagine this isn't a problem in real-world usage of this code as a TPM resource manager will be present, but I don't know for sure and it is annoying for testing purposes.

I can certainly add detection for C11 stdatomic and fall back to the compiler builtins.

Reply:
Yeah that would be great.